src/main/scala/workflow/AutoCacheRule.scala

```diff
@@ -0,0 +1,355 @@
+  }
+
+  def cacheMem(caches: Set[Int], profiles: Map[Int, Profile]): Long = {
+    // Must do a toSeq here otherwise the map may merge equal memories
+    caches.toSeq.map(i => profiles.getOrElse(i, Profile(0, 0)).mem).sum
+  }
+
+  /**
+   * Returns true iff there is still an uncached node whose output is used > once, that would fit in memory
+   * if cached
+   */
+  def stillRoom(caches: Set[Int], runs: Map[Int, Int], profiles: Map[Int, Profile], spaceLeft: Long): Boolean = {
+    runs.exists { i =>
+      (i._2 > 1) &&
+        (!caches.contains(i._1)) &&
+        (profiles.getOrElse(i._1, Profile(0, 0)).mem < spaceLeft)
+    }
+  }
+
+  def selectNext(
+    pipe: Seq[Instruction],
+    profiles: Map[Int, Profile],
+    cached: Set[Int],
+    runs: Map[Int, Int],
+    spaceLeft: Long
+  ): Int = {
+    //Get the uncached node which fits that maximizes savings in runtime.
+    pipe.indices.filter(i =>
+      !cached(i) &&
+        profiles.getOrElse(i, Profile(0, 0)).mem < spaceLeft &&
+        runs(i) > 1)
+      .minBy(i => estimateCachedRunTime(pipe, cached + i, profiles))
+  }
+
+  def greedyCache(
+    instructions: Seq[Instruction],
+    profiles: Map[Int, Profile],
+    maxMem: Long
+  ): Seq[Instruction] = {
+    val nodeWeights = getNodeWeights(instructions)
+
+    var cached = instructions.indices.filter { instructions(_) match {
+      case _: EstimatorNode => true
+      case _: TransformerNode =>  true
+      case _: EstimatorFitNode => true
+      case _: SourceNode => false
+      case _: TransformerApplyNode => false
+    }}.toSet
+
+    var usedMem = cacheMem(cached, profiles)
+    var runs = getRuns(instructions, cached, nodeWeights)
+    while (usedMem < maxMem && stillRoom(cached, runs, profiles, maxMem - usedMem)) {
+      cached = cached + selectNext(instructions, profiles, cached, runs, maxMem - usedMem)
+      runs = getRuns(instructions, cached, nodeWeights)
+      usedMem = cacheMem(cached, profiles)
+    }
+
+    makeCachedPipeline(instructions, cached -- WorkflowUtils.getChildren(Pipeline.SOURCE, instructions))
+  }
+
+  override def apply[A, B](plan: Pipeline[A, B]): Pipeline[A, B] = {
+    val instructions = WorkflowUtils.pipelineToInstructions(plan)
+
+    WorkflowUtils.instructionsToPipeline(cachingMode match {
+      case NaiveCache => naiveCache(instructions)
+      case GreedyCache(maxMem, profileScales, numProfileTrials) => {
+        val profiles = profileInstructions(instructions, profileScales, numProfileTrials)
+        greedyCache(instructions, profiles, maxMem)
+      }
+    })
+  }
+}
+
+object AutoCacheRule {
+  sealed trait CachingStrategy
+  /**
+   * NaiveCache is the optimal caching strategy assuming infinite memory and zero caching overhead
+   * (cache at every node with > 1 direct output dependency edge during training)
+   */
+  case object NaiveCache extends CachingStrategy
+  case class GreedyCache(
+    maxMem: Long,
+    profileScales: Seq[Long] = Seq(500, 1000),
```

Comment:
So - let's consider the VOC case - at 5000 examples the time spent profiling would be 30% of the total runtime in the best case, which seems awfully high. I *think* the way I had it before was as a function of the number of partitions (2 or 4 examples per partition, for example). I don't think we want users to have to explicitly choose this parameter (though it should be configurable) - instead we should pick something that we think would work and not slow things down too much in most cases. We have a few workloads that we can think about in this context, right? E.G. CIFAR is going to be 50k tiny images spread in nummachines*numcores partitions in the usual case, Amazon is 65m reviews spread across a similar number of partitions, TIMIT is 2m examples, and Solar Flares is some smaller number of images. How would each of these situations behave?